### ISSUE TITLE
Analysis of how implementations handle the required escaping in RegExp#source

### ISSUE THREAD

**Issue by claudepache** [tc39/ecma262#578]
This is a followup of https://bugs.ecmascript.org/show_bug.cgi?id=1470

I’ve made a first rapid analysis of how major web browsers implement the not-exactly-specified Step 2 of [EscapeRegExpPattern](https://tc39.github.io/ecma262/#sec-escaperegexppattern). Recall that, for a regexp `rx`, we have approximatively `rx.source = EscapeRegExpPattern(rx.[[OriginalSource]])`. That transformation must not change the semantics of the pattern, but is required in order that
```js
    eval("/" + rx.source + "/" + rx.flags)
```
produces a functionally equivalent regexp as `rx`.

Analysing the [grammar that is used to determine the limits of a regexp literal](https://tc39.github.io/ecma262/#sec-literals-regular-expression-literals), one can show that it suffices to:
* escape the four [line terminators](https://tc39.github.io/ecma262/#sec-line-terminators) in all positions;
* escape the character `/` outside [_RegularExpressionClass_](https://tc39.github.io/ecma262/#prod-RegularExpressionClass); and
* hack the empty pattern.

(Note that, although `/*` is parsed as a beginning of multiline-comment rather of a regular expression, this is nonproblematic because a regexp cannot ever begin with `*`.)

The transformations used by the major browsers are detailed below, except that the line terminators are currently not escaped by Chrome ([V8 Issue 1982](https://bugs.chromium.org/p/v8/issues/detail?id=1982)).

Original source         |  Transformed into
-----------------------|---------------
`<LF>`<br>`\<LF>`   |    `\n`
`<CR>`<br>`\<CR>`  |    `\r`
`<LS>`<br>`\<LS>`   |    `\u2028`
`<PS>`<br>`\<PS>`  |    `\u2029`
`/` (outside [_RegularExpressionClass_](https://tc39.github.io/ecma262/#prod-RegularExpressionClass)) | `\/`
`/` (inside [_RegularExpressionClass_](https://tc39.github.io/ecma262/#prod-RegularExpressionClass))  |  `/` (Firefox, Safari)<br>`\/` (Chrome, Edge)
empty pattern  |  `(?:)`

It does not seems to me that implementations perform other transformations, but that needs confirmation.

----
In conclusion, the only major difference between implementations seems to be whether `/` is escaped everywhere or only outside _RegularExpressionClass_.

**Comment by claudepache** [comment]
Personally, I am for not escaping `/` inside _RegularExpressionClass_, because that has the property of preserving exactly the source text when it originated from a regexp literal.
